### DISCUSSION TITLE
CORS Issue in Nextjs

### DISCUSSION THREAD

**Discussion by Revaycolizer** [vercel/next.js#56670] [Help] [answered]
### Summary

Recently I built a fullstack app with Nextjs but when I tried consuming the API's created within my Nextjs app I got this error 
![flutterm](https://github.com/vercel/next.js/assets/109684996/f11580f9-4b46-4044-8fff-fa8cd544e11e)

### Additional information

```js
my code snippets 

export async function POST(req: NextRequest, res: NextResponse) {

const response = {
      message: "Authenticated!",
    };
  
    const seralized = serialize(COOKIE_NAME, JSON.stringify(userData), {
      httpOnly: true,
      secure: process.env.NODE_ENV === "production",
      sameSite: "strict",
      maxAge: MAX_AGE,
      path: "/",
    });
  
    return new Response(JSON.stringify(response), {
      status: 200,
      headers: { 
        'Access-Control-Allow-Origin': '*',
      'Access-Control-Allow-Methods': 'GET, POST, PUT, DELETE, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type, Authorization',
      "Content-Type": "application/json" ,
        "Set-Cookie": seralized 
      },
    });

}

```
```


### Example

https://github.com/Revaycolizer/minimal-repo/blob/main/app/api/login/route.ts

**Comment by icyJoseph** [comment]
So you are running your app in localhost:11204 and try to reach the `api/login` endpoint from another domain?

You probably need to provide an options call. Refer to this response: https://github.com/vercel/next.js/discussions/52933#discussioncomment-6498462

**Reply by Revaycolizer** [reply]
So still I'll be able to use post if I use options?

**Reply by icyJoseph** [reply]
You need both methods. One answers the preflight request the browser uses to check for CORS, the other one handles the actual POST request.

**Reply by Revaycolizer** [reply]
> So you are running your app in localhost:11204 and try to reach the `api/login` endpoint from another domain?
> 
> You probably need to provide an options call. Refer to this response: https://github.com/vercel/next.js/discussions/52933#discussioncomment-6498462

I updated my codes you can take a look but added options still did not solve it

**Reply by icyJoseph** [reply, marked as answer]
But you only copied the Options method. Read the OP on the other link, you need to set the proper Cors headers. OP did it on middleware, but you should just do it in the OPTIONS request handler

**Reply by Revaycolizer** [reply]
> But you only copied the Options method. Read the OP on the other link, you need to set the proper Cors headers. OP did it on middleware, but you should just do it in the OPTIONS request handler 

Thanks now everything works just fine

**Reply by icyJoseph** [reply]
Mmm could you try to explain in more detail, step by step, what is it you are trying to do?

**Reply by Revaycolizer** [reply]
> Mmm could you try to explain in more detail, step by step, what is it you are trying to do?

Now everything works I just updated on the wrong project that's what keep causing the error but this time everything works fine thanks

**Reply by icyJoseph** [reply]
Nice! Don't forget to mark as answered! :)

**Reply by Revaycolizer** [reply]
> Nice! Don't forget to mark as answered! :)

Already
